### PR TITLE
Add diffGrad optimizer

### PR DESCRIPTION
**Pull request by suvojit-0x55aa**
# What is this ?
An implementation of the `diffGrad` optimizer as introduced in the [diffGrad: An Optimization Method for Convolutional Neural Networks](https://arxiv.org/abs/1909.11015) paper.

Resolves issue #27052

Adapted from author's [original implementation](https://github.com/shivram1987/diffGrad/blob/master/diffGrad.py) as requested by the authors.

### CODE REVIEW

**Comment by suvojit-0x55aa** [comment]
I noticed there are test for the `optim` module. However I'm having difficulty understanding what it's doing, is there any docs that I can reference ?

**Comment by gtamba** [comment]
Hey just my two cents, you should link back/attribute (at least in the PR header so that the dev team can review your reference) that you adapted the code from the author's online (repository)[https://github.com/shivram1987/diffGrad/blob/master/diffGrad.py].

**Comment by suvojit-0x55aa** [comment]
@gtamba Thanks for pointing that out, changed accordingly.

**Comment by soumith** [comment]
We don't think this method should be in pytorch core, as opposed to your own personal repository or something like https://github.com/pytorch/contrib , at least not yet in time.

Our reservation is that we want to include methods that the community uses as a standard, or else the code maintenance problem balloons up for us.
We do show discretion based on what the paper shows as evidence, for example BatchNorm was included within weeks of it's publish date (in Torch).
In terms of rejected methods, we've rejected newly minted papers such as Swish ( #3260 , #3182 ), Yellowfin ( https://github.com/pytorch/pytorch/issues/1960 ) and many others, and rightly so, these haven't become standardized in the community (like LSTM / Transformer / BatchNorm).

If you have a differing opinion, let us know why, and we can re-think.

tl;dr: The paper doesn't show evidence that makes it a method that has obvious long-term success. If the paper does have long-term success in the field we will include it

To save yourself time next time, please file an issue proposing what you plan to contribute, rather than doing all the work before-hand.
We write that at the top of the CONTRIBUTING GUIDE here: https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#contributing-to-pytorch
